tests/behavior/transforms/remap.toml

```diff
@@ -201,6 +201,63 @@
     [[tests.outputs.conditions]]
       "a.b\\.c.equals" = "baz"
 
+[transforms.remap_infallible_assignment]
+  inputs = []
+  type = "remap"
+  source = """
+    .nope, .err1 = parse_json("{ INVALID }")
+    .ok, .err2 = parse_json("{ \\"foo\\": true }")
+  """
+[[tests]]
+  name = "remap_infallible_assignment"
+  [tests.input]
+    insert_at = "remap_infallible_assignment"
+    type = "raw"
+    value = ""
+  [[tests.outputs]]
+    extract_from = "remap_infallible_assignment"
+    [[tests.outputs.conditions]]
+      "nope.equals" = "<null>"
+      "err1.equals" = "function call error: unable to parse json: key must be a string at line 1 column 3"
+      "ok.foo.equals" = true
+      "err2.equals" = "<null>"
+
+[transforms.remap_error_coalesce_operator]
+  inputs = []
+  type = "remap"
+  source = """
+    .val1 = parse_json("{ INVALID }") ?? "nope"
+    .val2 = parse_json("true") ?? "nope"
+  """
+[[tests]]
+  name = "remap_error_coalesce_operator"
+  [tests.input]
+    insert_at = "remap_error_coalesce_operator"
+    type = "raw"
+    value = ""
+  [[tests.outputs]]
+    extract_from = "remap_error_coalesce_operator"
+    [[tests.outputs.conditions]]
+      "val1.equals" = "nope"
+      "val2.equals" = true
+
+[transforms.remap_bang_function]
+  inputs = []
+  type = "remap"
+  source = """
```

Comment:
Presumably here if we did say `.a = 2`, even though the following line would fail, `a` is still getting set?

We should probably look into that.

Reply:
Correct.

This has been the case since the first version of VRL. @jszwedko has mentioned this several times as well. We would need some way to keep track of changes VRL makes to the object to roll them back if execution is aborted.

It's not something we support right now, and at a first glance also isn't trivial to add (it requires some rejigging of the internals). Mind opening a tracking issue for this @FungusHumungus?

Reply:
I opened https://github.com/timberio/vector/issues/6025 to track this.